nixos/modules/image/gpt.nix

```diff
@@ -0,0 +1,199 @@
+# This module exposes options to build a disk image with a GUID Partition Table
+# (GPT). It uses systemd-repart to build the image.
+
+{ config, pkgs, lib, utils, ... }:
+
+let
+  cfg = config.image;
+
+  partitionOptions = {
+    options = {
+      storePaths = lib.mkOption {
+        type = with lib.types; listOf path;
+        default = [ ];
+        description = lib.mdDoc "The store paths to inlcude in the partition.";
+      };
+
+      stripStorePaths = lib.mkOption {
+        type = lib.types.bool;
+        default = false;
+        description = lib.mdDoc ''
+          Whether to strip `/nix/store/` from the store paths. This is useful
+          when you want to build a partition that only contains store paths and
+          is mounted under `/nix/store`.
+        '';
+      };
+
+      contents = lib.mkOption {
+        type = with lib.types; attrsOf (submodule {
+          options = {
+            source = lib.mkOption {
+              type = types.path;
+              description = lib.mdDoc "Path of the source file.";
+            };
+          };
+        });
+        default = { };
+        example = lib.literalExpression '' {
+          "/EFI/BOOT/BOOTX64.EFI".source =
+            "''${pkgs.systemd}/lib/systemd/boot/efi/systemd-bootx64.efi";
+
+          "/loader/entries/nixos.conf".source = systemdBootEntry;
+        }
+        '';
+        description = lib.mdDoc "The contents to end up in the filesystem image.";
+      };
+
+      repartConfig = lib.mkOption {
+        type = with lib.types; attrsOf (oneOf [ str int bool ]);
+        example = {
+          Type = "home";
+          SizeMinBytes = "512M";
+          SizeMaxBytes = "2G";
+        };
+        description = lib.mdDoc ''
+          Specify the repart options for a partiton as a structural setting.
+          See <https://www.freedesktop.org/software/systemd/man/repart.d.html>
+          for all available options.
+        '';
+      };
+    };
+  };
+in
+{
+  options.image = {
```

Comment:
Not a fan of taking the namespace like this.

Reply:
This should make it clear this is using systemd-repart somehow.

Reply:
Mhmm does it need to encode systemd-repart though? Seems like a leaky abstraction that is not ncessary. But I can see that `options.image` is very fundamental. Would `options.diskImage` be better or what do you have in mind?

Reply:
I mean, if we do have other ways to build images, you will have to accommodate for them.
So except if you are ready to accept complicated discussions to compromise on the structure of this module, I don't think it's a good idea to have a global `image` module.

It would make more sense for me to have `systemd.repartImage.xyz`, it makes it clear this is a repart-based abstraction. The leaky abstraction is necessary because you have `repartConfig` in the whole thing.

Probably, we could have a `image.backend = "repart";` and `image.*` which have sensible options for general purpose and a proper abstraction, but I feel like it's premature?

Reply:
So, what about  `options.image.repart`? That leaves plenty of space for other backends and my current abstraction is leaky anyways already because it exposes `repartConfig`.

Yeah, you're right, let's keep this specific instead of generic for now.

Reply:
`image.repart` sounds good to me. :)